flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java

```diff
@@ -565,20 +563,37 @@ private void startTriggeringCheckpoint(CheckpointTriggerRequest request) {
                                                     request.props,
                                                     checkpointInfo.f0,
                                                     request.isPeriodic,
-                                                    checkpointInfo.f1.checkpointId,
-                                                    checkpointInfo.f1.checkpointStorageLocation,
+                                                    checkpointInfo.f1,
                                                     request.getOnCompletionFuture()),
                                     timer);
 
             final CompletableFuture<?> coordinatorCheckpointsComplete =
-                    pendingCheckpointCompletableFuture.thenComposeAsync(
-                            (pendingCheckpoint) ->
-                                    OperatorCoordinatorCheckpoints
-                                            .triggerAndAcknowledgeAllCoordinatorCheckpointsWithCompletion(
-                                                    coordinatorsToCheckpoint,
-                                                    pendingCheckpoint,
-                                                    timer),
-                            timer);
+                    pendingCheckpointCompletableFuture
+                            .thenApplyAsync(
+                                    pendingCheckpoint -> {
+                                        try {
+                                            CheckpointStorageLocation checkpointStorageLocation =
+                                                    initializeCheckpointLocation(
+                                                            pendingCheckpoint.getCheckpointID(),
+                                                            request.props,
+                                                            request.externalSavepointLocation,
+                                                            initializeBaseLocations);
+                                            pendingCheckpoint.setCheckpointTargetLocation(
```

Comment:
As I see now, according to our thread model here. It is better to avoid the volatile inside pendingCheckpoint. You can move setting the StorageLocation under the lock and to `timer` thread. So in this case you can move `pendingCheckpoint.setCheckpointTargetLocation` to next `thenComposeAsync` or create new small one. And use `synchronization(lock)`around the setting targetLocation. And then you can remove volatile for both fields(targetLocation, disposed).
I know it is not fully obvious but in parallel to your PR there is the task for moving all calls to JobManagerThread that make all execution here one threaded and it will allow avoiding synchronization at all. It is why it doesn't make sense to add new volatile in this PR.